### PR TITLE
Add templated generation of models

### PR DESCRIPTION
**Pull request by ttim**
I noticed we have a lot of duplication across different models, most recently with trt-llm. Some of the truss model configurations are the same apart from tiny parts of config where hugging face repo id is specified.
Eventually these models diverge, and this gets very hard to support.

One of the pragmatic ways to tackle is to support templated model configurations in `truss-examples` repo. There are "base" truss configurations which are located under `templates` directory, and models configurations at `templates/generate.yaml`.

Each configuration has 5 components: destination folder, `template` it's based on, files to ignore in destination folder (such as `README.md`) , truss config and Jinja template variables.

`generate.py` script located next to templates can be either run without arguments (and then it generates all of the models) or with `--only_check`  argument and then it fails if newly generated models differ from ones on disk. We use `--only_check` for CI action.

Currently I've added only `trt-llm` model, but I'll add `deepspeed` and `vllm` ones as well.

### CODE REVIEW

File: templates/generate.py [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+
+def process(dst: Path, templates: Path, generate: Generate, only_check: bool):
+    logging.info(f"processing {str(dst)}")
+    with tempfile.TemporaryDirectory() as temp_dir:
+        # copy template
+        for filename in os.listdir(templates / generate.based_on):
+            source_file = templates / generate.based_on / filename
+            destination_file = Path(temp_dir) / filename
+
+            if os.path.isfile(source_file):
```
Comment (bolasim) [inline templates/generate.py:35]: nit: best to use `pathlib.Path` functions over `os.path`. Is this case `source_file.is_file()`. Later you can use `source_file.read_text/write_text` etc

**Review by bolasim — COMMENTED** [review]
overall looks fine. I am curious how this interplays with the CI. Can you add one of the generated examples to the CI to ensure that everything runs smoothly? I'll leave it to @squidarth to approve.

**Comment by ttim** [comment]
@bolasim with this PR `mistral/mistral-7b-trt-llm` and `llama/llama-2-7b-trt-llm` are being generated from `templates/trt-llm`

**Review by squidarth — APPROVED** [review]
some small coments, otherwise lgtm

File: templates/trt-llm/data/.gitattributes [hunk 2]
```diff
@@ -0,0 +1,37 @@
+*.7z filter=lfs diff=lfs merge=lfs -text
```
[the remark this answers is not in the document]
Comment (ttim) [reply templates/trt-llm/data/.gitattributes:1]: Was in original trt-llm model

**Comment by squidarth** [comment]
@ttim  wrt. CI, would be good to add the llama & mistral trt trusses to the ci.yaml file

File: templates/README.md [hunk 3]
```diff
@@ -0,0 +1,2 @@
+# Truss templates for different model backends
```
[the remark this answers is not in the document]
Comment (ttim) [reply templates/README.md:1]: Done

File: templates/generate.py [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+            shutil.copy2(dst / ignored, generated / ignored)
+        # apply config changes
+        template_config = yaml.safe_load((generated / "config.yaml").read_text())
+        merged_config = merge_configs(template_config, generate.config)
+        (generated / "config.yaml").write_text(merged_config)
+        # apply replaces
+        for file, replace in generate.replaces.items():
+            content = (generated / file).read_text()
+            (generated / file).write_text(
+                content.replace(replace.from_str, replace.to_str)
```
[the remark this answers is not in the document]
Comment (ttim) [reply templates/generate.py:44]: Good idea, will do

**Comment by ttim** [comment]
@squidarth I've added pre commit hook which checks that all of the generated models are up-to-date. Currently this includes mistral and llama trt ones.

File: templates/generate.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (ttim) [reply templates/generate.py]: Done

**Pull request merged**